### ISSUE TITLE
Should it be possible to connect to a simulated local mobile reader?

### ISSUE THREAD

**Issue by mihaildu** [stripe/stripe-terminal-react-native#439]
**Describe the bug**
I'm trying to connect to a simulated local mobile reader and I'm getting this error
```
const {reader: connectedReader, error} = await connectLocalMobileReader({
  reader,
  locationId,
});

// error = [TypeError: undefined is not a function]
```

**To Reproduce**
Discover simulated local mobile readers & try to connect
```
const {error} = await discoverReaders({
  discoveryMethod: 'localMobile',
  simulated: true,
});
...
// discoveredReaders - I have simulated Stripe S700, wisePosE, Simulated Etna and verifoneP400 here
// connectToLocalMobileReader calls the above code
connectToLocalMobileReader(discoveredReaders[0])
``` 

**Expected behavior**
Not sure, I'm just curious if this should work or not with simulated local mobile readers

**Stripe Terminal React Native SDK version**

- ^0.0.1-beta.11

**Smartphone (please complete the following information):**

- Simulated iPhone 14 Pro with iOS 13.7

**Comment by billfinn-stripe** [comment]
Good question -- no, connecting to simulated local mobile readers is not supported. (In fact, currently, local mobile (aka, tap-to-pay) is not supported by ReactNative at all.)

I'll leave this issue open as a reminder to improve the error message and/or documentation.

**Comment by FaggioniHQ** [comment]
Hello @billfinn-stripe ,

I'm implementing tap-to-pay on a react native app using the iOS terminal SDK. The only issue that I found is that the react SDK uses the StripeTerminal version 2.13.0, and for supporting iPhone payments we need to use the StripeTerminal version 2.17.1. This leads to a conflict between the SDKs.

My question is, Are you planning to bump the react native SDK to use StripeTerminal 2.17.1? 

We support payments using Bluetooth readers with the react native SDK, we would like to keep both methods.

**Comment by FaggioniHQ** [comment]
Hello again @billfinn-stripe ,

I review the repo and noticed that you already working on it. That's perfect.

Any ETA for this? 

Thanks in advance

**Comment by billfinn-stripe** [comment]
Hi @FaggioniHQ -- no ETA yet. Thanks for flagging it as a need for you and your team. Our current plan is to bump the native versions to our 3.x.x beta versions before releasing a new RN SDK. We're relying on open source PRs to move this along, so we don't have any firm time commitments. If we haven't gotten any momentum for bumping to 3.x.x in the next two weeks, we'll release a RN version that is pointing to the latest generally available native SDK versions.

**Comment by FaggioniHQ** [comment]
Hello @billfinn-stripe ,

Thanks for the quick response. Sounds good.

Thanks for your help.

**Comment by mihaildu** [comment]
> Good question -- no, connecting to simulated local mobile readers is not supported. (In fact, currently, local mobile (aka, tap-to-pay) is not supported by ReactNative at all.)

Good to know, thank you for the answer!

> I'll leave this issue open as a reminder to improve the error message and/or documentation.

I think the documentation is pretty clear that it only works with iOS SDK, but after looking through the example app in the github repo where it seems like it might work, I decide to give it a try and hope for outdated documentation :)

I have a follow up question: to make Tap to Pay work with RN is it just about bumping StripeTerminal dep to 2.14+ (already done on main branch) and also adding an extra function to [StripeTerminalReactNative.swift](https://github.com/stripe/stripe-terminal-react-native/blob/main/ios/StripeTerminalReactNative.swift) called `connectLocalMobileReader` that should try to connect to a local mobile reader, similar to the example [here](https://github.com/stripe/stripe-terminal-ios/blob/master/Example/Example/ReaderDiscoveryViewController.swift#L160-L161)? I might be able to submit this PR if it's only those changes.

Also, just wanted to say I would also like this feature implemented in RN for my team. Thank you!

**Comment by mihaildu** [comment]
> I see that the native Android code already seems to support it. Did you happen to test if tap-to-pay already works for Android if you bump the underlying native SDK versions?

Unfortunately I didn't get a chance to test anything with Android, we don't plan to support Android until later down the line.

> Also, we'll want to add localmobile to DiscoveryMethod [here](https://github.com/stripe/stripe-terminal-react-native/blob/92aab094eb7438bdff8d77de34c34e4d330b7db5/src/types/Reader.ts#L29-L32).

Great, I created a fork [here](https://github.com/mihaildu/stripe-terminal-react-native) with the suggested changes which can be seen in this [PR](https://github.com/mihaildu/stripe-terminal-react-native/pull/1). I will give it a test when I have some time. Let me know if you want me to also open the PR against this repository!

> I confirmed internally, and we are working on adding support for tap-to-pay to RN, but I can't commit to a timeline.

Ok cool! Thanks for letting me know

**Comment by billfinn-stripe** [comment]
> Great, I created a fork [here](https://github.com/mihaildu/stripe-terminal-react-native) with the suggested changes whch can be seen in this [PR](https://github.com/mihaildu/stripe-terminal-react-native/pull/1). I will give it a test when I have some time. Let me know if you want me to also open the PR against this repository!

Great, thanks! We may also want to map more native errors to RN-friendly error strings (so we don't see "Unknown" for tap-to-pay specific errors that haven't been mapped yet) [0], but we can do that in a follow-up PR.

Let me know how testing goes, and, if it goes well, we'd welcome a PR to this repo too. Thanks!

[0] https://github.com/stripe/stripe-terminal-react-native/blob/61f9dc2d9c5f488f012ea2a016808ce57e35effa/ios/Errors.swift#L45

**Comment by mihaildu** [comment]
> Let me know how testing goes, and, if it goes well

Well I managed to get it to work up until the Tap to Pay on iPhone screen, where you get to choose an Apple ID. However when calling `collectPaymentMethod` it errors out
```
Error collecting the payment intent: {“code”: “InvalidRequiredParameter”, “message”: “A required parameter was invalid or missing.“}
```
I would like to debug this but the code for `Terminal.shared.collectPaymentMethod` is not open source afaik. I set all the required params when creating the payment intent and when collecting it. 
```
const {error, paymentIntent} = await createPaymentIntent({
  amount: 100,
  currency: 'usd',
  onBehalfOf: '<some-test-acct>',
  transferDataDestination: '<some-test-acct>',
  applicationFeeAmount: 20,
});
...
const {
  paymentIntent: collectedPaymentIntent,
  error,
} = await collectPaymentMethod({
  paymentIntentId: paymentIntent.id,
});
```

Maybe it's a dumb question, but does my Apple ID have to be from US? I tested using a device location from US, but I'm not physically from US.

> we'd welcome a PR to this repo too

I am more than happy to open a PR from my fork, but I will need a little bit of guidance. This is the first time I'm writing any Swift code or do any mobile development, so I have a few questions. If this is OK with you, I will open a PR :)

**Comment by billfinn-stripe** [comment]
Great, thanks. I left [a comment on the PR](https://github.com/mihaildu/stripe-terminal-react-native/pull/1/files#r1129913625). When creating Payment Intents on behalf of other accounts, you'll also need to connect to the reader on behalf of the other account:

https://stripe.dev/stripe-terminal-ios/docs/Classes/SCPLocalMobileConnectionConfiguration.html#/c:objc(cs)SCPLocalMobileConnectionConfiguration(py)onBehalfOf

> I am more than happy to open a PR from my fork, but I will need a little bit of guidance. This is the first time I'm writing any Swift code or do any mobile development, so I have a few questions. If this is OK with you, I will open a PR :)

Let's get it tested and working, and then we'd love a PR that includes the entire changeset. Thank you!

**Comment by mihaildu** [comment]
> Sounds good. I still have a hunch it's related to on_behalf_of

You are right, I forgot to add it when I call `connectLocalMobileReader` from RN. Now it all works!

I opened my PR here https://github.com/stripe/stripe-terminal-react-native/pull/441

Not related to this PR, I am also wondering if `createPaymentIntent` should accept a `captureMethod` param. I don't want to use `manual` and it would be nice if I can do all of this without backend support. Looking at [API docs](https://stripe.com/docs/api/payment_intents/create#create_payment_intent-capture_method) it says it has a default value of `automatic`, so something is explicitly setting it to `manual`

**Comment by mihaildu** [comment]
> Yep, definitely we should add that

Do you have an ETA for this?

> Would you mind adding a few notes in the PR description what manual testing you performed?

Sure, I'll add some description with the steps.

Awesome, thanks!

**Comment by chenzesam** [comment]
@billfinn-stripe Hi, I understand that you guys discussed publishing RN code for TTPOI two weeks ago. 
Are there any plans to release it recently, please?

**Comment by liran-co** [comment]
@billfinn-stripe Is there any update on this?

**Comment by tenacioustechie** [comment]
Any plans to release this? Very keen to start developing against this, otherwise we might need to start our own version off @mihaildu 's PR.

**Comment by gperl27** [comment]
Has anyone had success using the beta version that includes #441 with a real device? Is the beta only supposed to work in a simulated environment? No matter what, `discoverReaders` returns an empty array. 

At first I thought it was my provisioning profile and/or entitlements file, but I cross-checked that they are correct by installing the [stripe iOS example app](https://github.com/stripe/stripe-terminal-ios); here I was able to discover my device as a reader and initiate a TTPOI payment.

I also tried patching `@stripe/stripe-terminal-react-native`'s podspec to Stripe SDK `2.14.0` and the `latest` (2.19.1 as of today) version with the same results.

**Comment by chr-stripe** [comment]
Hi folks, thanks for reaching out to us about getting support for Tap to Pay in the React Native SDK. Right now we’re targeting a new release for end of May at the latest. We’ll update this thread as well once the next release lands.

**Comment by nazli-stripe** [comment]
Hi folks, thanks for your patience. We are planning on releasing the new version with Tap to Pay Android and iOS support by end of this week.

**Comment by nazli-stripe** [comment]
Tap to Pay support added in the latest release: https://github.com/stripe/stripe-terminal-react-native/releases/tag/v0.0.1-beta.12

**Comment by 592da** [comment]
I having this error as well, using beta.12

```
{code: InvalidRequiredParameter, message: A required parameter was invalid or missing.}
```

the payment intent is created in the server, important to note, works with WisePad3 with the same flow.
might relate to the fact my apple id is registered in the UK.

(the merchant is set up in US, might not have location settings on as I'm debugging it from the UK)
@nazli-stripe

**Comment by Joshandrews43** [comment]
Feedback for the stripe team @billfinn-stripe

1. When using `on_behalf_of`, no where in the docs does it say to use it during the `connectLocalMobileReader` call in your docs. It just throws the `A required parameter was invalid or missing. Collecting a PaymentMethod for a PaymentIntent `on_behalf_of` a Connected Account requires providing that Connected Account's ID in `SCPLocalMobileConnectionConfiguration` error, which had me looking at the collect payment method functions for a few hours. 

2. There is a chicken and egg problem I encountered where Apple requires a video to approve an app for tap-to-pay to allow local reader discoveries, but the entitlement they give you does not include ad-hoc builds, so you can only use the simulated version, which they did not accept. I had to reach out to apple support to get ad-hoc builds in development to work for local readers (I use expo & all builds are ad-hoc).

**Comment by ivanmilosavljevic** [comment]
Feedback for the stripe team @billfinn-stripe 

Greetings from the IT team -  **Belgrade, Serbia**. 
We are developing the Lobby Kiosk Android App in React Native.

Our product/app offers all kinds of services for users to book. Our Lobby Kiosk is from 22in-I-Series-4 and we found it on the list of registered device Kiosk -> Elo Touch Solutions -> 22in-I-Series-4, developing in Android 12, SDK 31.

Since the last step of the Booking process is the Checkout with Payment option, and we have implemented **stripe-terminal-react-native**, we are experiencing troubles with our payment process.

Here are the steps in which Stripe Terminal is implemented:

- In our file with routes, we are wrapping our Stack. Screen with **StripeTerminalProvider** in which we are passing token provider
 - In our first child component (Home page), we are initializing stripe integration, as described in the documentation

- On the Checkout Page, we trigger the process of discovering readers:

1. Firstly, we are getting permissions (PermissionsAndroid.PERMISSIONS.ACCESS_FINE_LOCATION, PermissionsAndroid.PERMISSIONS.BLUETOOTH_SCAN, PermissionsAndroid.PERMISSIONS.BLUETOOTH_CONNECT, PermissionsAndroid.PERMISSIONS.ACCESS_COARSE_LOCATION)

2.  Triggering Discover Reader via 

 ```
const { error } = await discoverReaders({
          discoveryMethod: "localMobile"
        }); 
```

method from **useStripeTerminal** hook and in the **onUpdateDiscoveredReaders** method we are proceeding to connect with the reader (first discovered reader).

Note: In this step, we are getting only one “Cots” Reader in the list of discovered readers and this is the same with simulated false or true - testing it on the Lobby Kiosk device. Location ID we are providing from our BE and it is the location ID from the Stripe Terminal Admin/dashboard.

 ```
const { reader, error } = await connectLocalMobileReader({
          reader: discoveredReaders[0],
          locationId: locationId
        });
```


 3. Payment process - After this, we are starting the process of payment in the following steps from the FE side:

    1. fetching payment intent from our BE  
    2. retreiving payment intent retrievePaymentIntent method from useStripeTerminal
    3. collecting payment intent  collectPaymentMethod method from useStripeTerminal
    4. confim payment intent confirmPaymentIntent method from useStripeTerminal


Analyzing state currently, we can’t pass by this local mobile discovery or connect with the reader, here are the logs from our Elo Lobby Kiosk logs:

```
2025-03-18 14:59:09.784 15744 15772 D StripeTerminal: class=Terminal message="Application version info" minSdkVersion=26 targetSdkVersion=34 compileSdkVersion=35

2025-03-18 14:59:10.270 15744 15772 I StripeTerminal: class=Terminal message="localmobile module detected"

2025-03-18 14:59:10.873 15744 15772 D StripeTerminal: class=Terminal app_language=KOTLIN
2025-03-18 14:59:10.875 15744 15772 D StripeTerminal: class=LocationHandler message=startListening

2025-03-18 14:59:11.139 15744 15772 D StripeTerminal: class=ConnectionTokenRepository message=refreshToken

2025-03-18 14:59:11.203 15744 15863 I StripeTerminal: class=DefaultStripeConnectivityRepository message="continuing health checks, sdk embedded status: false"
2025-03-18 14:59:11.217 15744 15868 D StripeTerminal: class=ProxyOfflineListener message=onOfflineStatusChange
2025-03-18 14:59:11.220 15744 15772 D StripeTerminal: class=DefaultOfflineTraceManager message="Start forwarding" isAlreadyForwarding=null
2025-03-18 14:59:11.221 15744 15861 D StripeTerminal: class=DefaultOfflineTraceManager message="Fetch connection token attempt: 0"
2025-03-18 14:59:11.223 15744 15771 I ReactNativeJS: '[Stripe terminal]: didChangeOfflineStatus', { reader: undefined,
2025-03-18 14:59:11.226 15744 15861 D StripeTerminal: class=DefaultOfflineTraceManager message="Forwarding started" hasAccount=false hasNetwork=true
2025-03-18 14:59:11.448 15744 15869 D StripeTerminal: class=ProxyOfflineListener message=onOfflineStatusChange
2025-03-18 14:59:11.450 15744 15771 I ReactNativeJS: '[Stripe terminal]: didChangeOfflineStatus', { reader: undefined,
2025-03-18 14:59:11.875 15744 15772 D StripeTerminal: class=ConnectionTokenRepository message=fetchConnectionToken.onSuccess
2025-03-18 14:59:11.881 15744 15861 D StripeTerminal: class=DefaultOfflineTraceManager message="Retrieving account details."
025-03-18 14:59:12.173 15744 15861 D StripeTerminal: class=DefaultOfflineTraceManager message="Forwarding started" hasAccount=true hasNetwork=true
2025-03-18 14:59:43.590 15744 15772 D StripeTerminal: class=Terminal message=discoverReaders
2025-03-18 14:59:43.591 15744 15772 I StripeTerminal: class=TerminalSession message=enqueueOperation operation=DiscoverReadersOperation
2025-03-18 14:59:43.596 15744 16332 D StripeTerminal: class=ProxyAdapter message="Switching from null to CotsAdapter"
2025-03-18 14:59:43.602 15744 15866 D StripeTerminal: class=AidlRpcClient message=bindToService
2025-03-18 14:59:43.606 15744 15866 D StripeTerminal: class=AidlRpcClient message="Context::bindService completed" success=true
2025-03-18 14:59:45.361 15744 15744 D StripeTerminal: class=AidlServiceConnection message=onServiceConnected() connectionOwner=CotsClient
2025-03-18 14:59:45.363 15744 16332 D StripeTerminal: class=AidlRpcClient message=makeRequest service=COTS_SERVICE method=DISCOVER requestId=0
```

Important note: When we are using a simulated Bluetooth reader we don’t experience these issues and we can go through with payment. 

For discovering we use:
```
  const { reader, error } = await connectBluetoothReader({
          reader: discoveredReaders[0],
          // Since the simulated reader is not associated with a real location, we recommend
          // specifying its existing mock location.
          locationId: discoveredReaders[0].locationId
        });
```
and for connecting 
```
    const { error } = await discoverReaders({
          discoveryMethod: "bluetoothScan",
          simulated: true // or false depending on your environment
        });
```



Since Tap to Pay is crucial for our Lobby kiosk device, and our kiosk supports NFC payment, we are looking for a way to connect and implement this in our payment system.

Please walk us through the requirements so this can be possible.


Here are dependencies: 
```
"dependencies": {
    "@react-native-async-storage/async-storage": "^2.1.0",
    "@react-native-community/datetimepicker": "^8.3.0",
    "@react-native-community/geolocation": "^3.4.0",
    "@react-native-community/hooks": "^3.0.0",
    "@react-native-community/netinfo": "^11.4.1",
    "@react-native-picker/picker": "^2.11.0",
    "@react-navigation/native": "^7.0.14",
    "@react-navigation/native-stack": "^7.2.0",
    "@stripe/stripe-terminal-react-native": "^0.0.1-beta.23",
    "axios": "^1.7.9",
    "date-fns": "^4.1.0",
    "i18next": "^24.2.2",
    "react": "18.3.1",
    "react-i18next": "^15.4.0",
    "react-native": "0.76.5",
    "react-native-asset": "^2.1.1",
    "react-native-config": "^1.5.3",
    "react-native-device-info": "^14.0.2",
    "react-native-fs": "^2.20.0",
    "react-native-gesture-handler": "^2.22.1",
    "react-native-localize": "^3.4.1",
    "react-native-maps": "^1.20.1",
    "react-native-permissions": "^5.2.5",
    "react-native-reanimated": "^3.16.7",
    "react-native-reanimated-carousel": "^3.5.1",
    "react-native-safe-area-context": "^5.0.0",
    "react-native-screens": "^4.3.0",
    "react-native-svg": "^15.11.1",
    "react-native-svg-transformer": "^1.5.0",
    "react-native-vector-icons": "^10.2.0",
    "styled-components": "^6.1.15"
  },
  "devDependencies": {
    "@babel/core": "^7.25.2",
    "@babel/preset-env": "^7.25.3",
    "@babel/runtime": "^7.25.0",
    "@react-native-community/cli": "15.0.1",
    "@react-native-community/cli-platform-android": "15.0.1",
    "@react-native-community/cli-platform-ios": "15.0.1",
    "@react-native/babel-preset": "0.76.5",
    "@react-native/eslint-config": "0.76.5",
    "@react-native/metro-config": "0.76.5",
    "@react-native/typescript-config": "0.76.5",
    "@types/react": "^18.2.6",
    "@types/react-native-vector-icons": "^6.4.18",
    "@types/react-test-renderer": "^18.0.0",
    "@types/styled-components": "^5.1.34",
    "babel-jest": "^29.6.3",
    "eslint": "^8.19.0",
    "jest": "^29.6.3",
    "prettier": "^3.4.2",
    "react-redux": "^8.1.2",
    "react-test-renderer": "18.3.1",
    "reactotron-react-native": "^5.1.12",
    "redux": "^4.2.1",
    "redux-thunk": "^2.4.2",
    "typescript": "5.0.4"
  }
```
 Thank you in advance for your assistance and we are looking forward to hearing from you.
